labs.md

```diff
@@ -0,0 +1,181 @@
+Keep all implementation files in the same folder and in package `main`
+Test against the `Storer` interface and run in [suite](https://godoc.org/github.com/stretchr/testify/suite) with both implementations
+
+# Lab 7 - Errors
+
+Create an executable go program in directory `07_errors/USERNAME`
+Copy the CRUD puppy from upstream master `06_puppy/USERNAME`
+Add a custom error type `Error` with fields `Message` and `Code`
+Extend the `Storer` interface for all methods to also return `error`
+Create errors for:
+
+        - Value < 0
+        - ID not found in Read, Update and Delete
+
+Add locking for proper use of sync.Map
+Bonus points: Add a third `Storer` implementation using [leveldb](https://github.com/syndtr/goleveldb)
+
+# Lab 8 - Project Layout
+
+Copy the CRUD puppy from upstream master `07_errors/USERNAME`
+Create directory `08_project/USERNAME` containing
+
+```
+├── README.md
+├── go.mod
+├── go.sum
+├── pkg
+│   └── USERNAME-puppy
+│       ├── types.go
+│       ├── types_test.go
+│       ├── errors.go
+│       ├── errors_test.go
+│       └── store
+│           ├── storer.go
+│           └── .... store files and tests, e.g. mapstore.go
+├── cmd
+│   └── USERNAME-puppy-server
+│       └── main.go
+└── vendor
+```
+
+Add project introduction and how to build, run & test it to `README.md`
+
+# Lab 9 - JSON puppy
+
+Create directory `09_json/USERNAME` containing a copy of upstream master `08_project/USERNAME`
+Add JSON tags to puppy data type
+Test marshalling and unmarshalling using [require.jsoneq](https://godoc.org/github.com/stretchr/testify/require#JSONEq)
+Add command line flag `-d FILE` with long form `--data FILE` using [kingpin.v2](https://godoc.org/gopkg.in/alecthomas/kingpin.v2)
+FILE should contain an array of puppies in JSON format. Parse this file and store its contents.
+
+# Lab 10 - Puppy REST
+
+Create directory `10_rest/USERNAME` containing a copy of upstream master `09_json/USERNAME`
+Add file `pkg/puppy/rest.go` implementing:
+
+```
+GET    /api/puppy/{id}
+POST   /api/puppy/          Payload: Puppy JSON without ID
+PUT    /api/puppy/{id}      Payload: Puppy JSON without ID
+DELETE /api/puppy/{id}
+```
+
+Use [net/http/httptest](https://golang.org/pkg/net/http/httptest/) for testing
+Add flag `-p PORT` with long flag `--port PORT` to command line flags
+Add flag `-s STORE` with long flag `--store STORE` with accepted values:
+
+```
+map, sync, db
+```
+Document the API in README.md
+
+# Lab 11 - Puppy Notifications
+
+Create directory `11_notify/USERNAME` containing a copy of upstream master `10_rest/USERNAME`
+Create `cmd/lostpuppy-service/main.go` running single endpoint:
+```
+POST   /api/lostpuppy/          Payload: { id: PUPPY_ID }
+This stubbed endpoint returns with 2 second delay:
+HTTP status 201 for even IDs
+HTTP status 500 for odd IDs
+```
+Update Puppy Delete method to notify lostpuppy-service in a goroutine and log response code asynchronously.
```

Comment:
There is no newline at the end of the file. Every line (even the last) should be terminated with a new line.

Unfortunately, it looks (to me) that GitHub is no longer showing this - it used to with a red marker. Please configure your editor to always include a newline at the end of every line, including the last.